bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/WriteEntryProcessor.java

```diff
@@ -114,9 +115,10 @@ public void writeComplete(int rc, long ledgerId, long entryId,
             requestProcessor.addEntryStats.registerFailedEvent(MathUtils.elapsedNanos(startTimeNanos),
                     TimeUnit.NANOSECONDS);
         }
-        sendResponse(rc,
-                     ResponseBuilder.buildAddResponse(request),
-                     requestProcessor.addRequestStats);
+
+        Response response = ResponseBuilder.buildAddResponse(request);
+        sendResponse(rc, response, requestProcessor.addRequestStats);
```

Comment:
sendResponse calls channel.writeAndFlush(response...). Are we sure that the writeAndFlush isn't keeping a reference to response somewhere?

Reply:
I believe that when the flush is called, the whole chain of handlers is invoked. Therefore, even if the channel is not writable, we should already have converted the add response into a buffer.